Final touch ups for the modernization effort

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This pull request is a final "clean up" for the project modernization effort. It includes the following:

- Integrate the latest from [cisagov/skeleton-docker].
- Update the dependabot configuration to check our [pipenv] dependencies.
- Bump the version of Python used from `3.11.1` to `3.11.2`.
- Bump the versions of the following packages:
  - [pip] from `22.3.1` to `23.0`
  - [setuptools] from `65.7.0` to `67.3.1`
  - [pipenv] from `2022.12.19` to `2023.2.4`
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass. I verified that the output of this prerelease matches the currently in production output.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

[cisagov/skeleton-docker]: https://github.com/cisagov/skeleton-docker
[pip]: https://pypi.org/project/pip/
[pipenv]: https://pypi.org/project/pipenv/
[setuptools]: https://pypi.org/project/setuptools/

jsf9k committed:
Specify the default group to run under in the Dockerfile

jsf9k committed:
Update wheel along with pip and setuptools

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add package write permission to workflow

When Dependabot creates a PR it requires this permission in order to
push Docker images to ghcr.io.

mcdonnnj committed:
Update dependabot configuration

Add another `pip` package ecosystem for the `src/` directory so the
`pipenv` configuration is seen and maintained by dependabot.

mcdonnnj committed:
Update Python package versions in the Dockerfile

- Bump pip from 22.3.1 to 23.0
- Bump setuptools from 65.7.0 to 67.3.1
- Bump pipenv from 2022.12.19 to 2023.2.4

mcdonnnj committed:
Update the `pipenv` lockfile

This was done by running `pipenv lock` in the `src/` directory.

mcdonnnj committed:
Correct the minor version tag in the README

This was missed when the version was bumped.

mcdonnnj committed:
Update pre-commit hooks

Update pre-commit hooks using `pre-commit autoupdate`. The `ansible-lint` hook
is intentionally held back to be updated independently to v6.

jmorrowomni committed:
Merge pull request #124 from cisagov/maintenance/update_pre-commit_hooks

Update `pre-commit` hooks

Commit:
Merge https://github.com/cisagov/skeleton-generic into lineage/skeleton

jsf9k committed:
Add a security label

jmorrowomni committed:
Merge pull request #125 from cisagov/improvement/add-security-label

Add a security label

Commit:
Merge https://github.com/cisagov/skeleton-generic into lineage/skeleton

jmorrowomni committed:
Merge pull request #149 from cisagov/lineage/skeleton

Lineage pull request for: skeleton

jmorrowomni committed:
Merge pull request #144 from cisagov/bugfix/add-package-permission

Add package write permission to workflow

jmorrowomni committed:
Merge pull request #143 from cisagov/improvement/update-wheel

Update wheel along with pip and setuptools

jmorrowomni committed:
Specify pip3

In this case it doesn't matter because we are starting from a
Python3-specific base container, but other projects that use this
skeleton may not be.  Specifying pip3 ensures that the Python 2
version of pip is not called by mistake.

jmorrowomni committed:
Merge pull request #142 from cisagov/improvement/specify-python3

Specify pip3

jmorrowomni committed:
Merge pull request #141 from cisagov/improvement/specify-group

Specify the default group to run under in the Dockerfile

dependabot[bot] committed:
Bump docker/build-push-action from 3 to 4

Bumps [docker/build-push-action](https://github.com/docker/build-push-action) from 3 to 4.
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/v3...v4)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

jmorrowomni committed:
Specify pip3 in the file where it got blown away by a rebase

jmorrowomni committed:
Merge pull request #151 from cisagov/dependabot/github_actions/docker/build-push-action-4

Bump docker/build-push-action from 3 to 4

jmorrowomni committed:
Merge pull request #154 from cisagov/improvement/update_pip3

Specify pip3 in the file where it got blown away by a rebase

mcdonnnj committed:
Merge github.com:cisagov/skeleton-docker into improvement/last_touchups

mcdonnnj committed:
Bump cisagov/mongo-db-from-config from 0.1.0 to 0.2.0

mcdonnnj committed:
Bump python from 3.11.1 to 3.11.2

This includes the Docker images used as well as the version specified
in the Pipfile.

mcdonnnj committed:
Update the `mypy` pre-commit hook configuration

Add some additional typing packges to improve analysis. Since this is a
Docker project it doesn't make sense to install packages into the local
Python environment for analysis.

mcdonnnj committed:
Bump version from 0.1.0-rc.1 to 0.1.0-rc.2

jsf9k commented on Dockerfile:
```diff
@@ -26,17 +26,17 @@ ENV VIRTUAL_ENV="${CISA_HOME}/.venv"
+    pip==23.0 \
+    setuptools==67.3.1 \
     wheel==0.38.4 \
   && python3 -m pip install --no-cache-dir --upgrade \
-    pipenv==2022.12.19 \
+    pipenv==2023.2.4 \
   # Manually create Python virtual environment for the final image
   && python3 -m venv ${VIRTUAL_ENV} \
   # Ensure the core Python packages are installed in the virtual environment
   && ${VIRTUAL_ENV}/bin/python3 -m pip install --no-cache-dir --upgrade \
-    pip==22.3.1 \
-    setuptools==65.7.0 \
+    pip==23.0 \
+    setuptools==67.3.1 \
```
Is it worth making variables for the `pip`, `pipenv`, `setuptools`, and `wheel` versions so that they can all be updated in one place?  This would also DRY out the code since the `pip`, `setuptools`, and `wheel` versions are currently being specified twice (lines 29-30 and 38-39).

dav3r replied:
I support this idea.

mcdonnnj replied:
My original rationale was thinking that obfuscating the version might hinder any dependabot analysis that might automatically detect it as a dependency. After thinking about it though I thought that this approach is probably better overall to both ensure the same version is installed consistently and potentially avoiding multiple dependabot PRs around the frequently updating packages (`pip`, `pipenv`, and `setuptools`). These package versions can always be updated whenever dependabot generates a new `python` Docker image PR. Please see [685be83](https://github.com/cisagov/client-cert-update/pull/25/commits/685be832be4309a6f2ebd43faba87dac2a0ee68e).

jsf9k reviewed (approved):
This looks great to me.  I made one suggestion, but even if you choose to ignore it this PR is still approved.

dav3r reviewed (approved):
Looks great - nice work! 👍

mcdonnnj committed:
Store package versions as Docker environment variables

Update the Docker configuration to use environment variables to house
the versions of pip, pipenv, setuptools, and wheel that are installed
directly in the Dockerfile. This will DRY out the configuration a bit
as well as ensuring that the same versions are used for packages that
are installed multiple times.

The environment variable name format is cribbed from the official
Python Docker image configuration:
https://github.com/docker-library/python/blob/0a56f04a07495e0ed61663c73bc0c2e175773ddc/3.11/alpine3.17/Dockerfile#L132-L135
I tested without the leading `PYTHON_` in the names and only observed a
failure for usage of `PIPENV_VERSION`. Following the above mentioned
standard seemed like a reasonable approach as a result.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r commented on Dockerfile:
```diff
@@ -16,6 +16,13 @@ ARG CISA_USER="cisa"
 
+
+# Versions of the Python packages installed directly
+ENV PYTHON_PIP_VERSION=23.0
+ENV PYTHON_SETUPTOOLS_VERSION=67.3.1
+ENV PYTHON_WHEEL_VERSION=0.38.4
+ENV PYTHON_PIPENV_VERSION=2023.2.4
```
Might as well alphabetize these suckas:
```suggestion
ENV PYTHON_PIP_VERSION=23.0
ENV PYTHON_PIPENV_VERSION=2023.2.4
ENV PYTHON_SETUPTOOLS_VERSION=67.3.1
ENV PYTHON_WHEEL_VERSION=0.38.4
```

mcdonnnj replied:
That's fair. I'd originally ordered them intentionally by order of use but considering how often we beat the drum about alphabetizing unless there's a technical reason we cannot do so it was a bit silly of me 😅

dav3r replied:
Haha, I had a feeling that was what you had been thinking!

jsf9k reviewed (approved):
_Approval intensifies!!!_

mcdonnnj committed:
Sort Python package versions in the Dockerfile

Since we regularly beat the drum about alphabetizing such things unless
there are technical constraints we should definitely be alphabetizing
these variables.

Co-authored-by: dav3r <daver@geekpad.com>
